exporter/mezmoexporter/exporter.go

```diff
@@ -0,0 +1,174 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mezmoexporter
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strings"
+	"sync"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/plog"
+)
+
+type mezmoExporter struct {
+	config   *Config
+	settings component.TelemetrySettings
+	client   *http.Client
+	wg       sync.WaitGroup
+}
+
+type MezmoLogLine struct {
+	Timestamp int64             `json:"timestamp"`
+	Line      string            `json:"line"`
+	App       string            `json:"app"`
+	Level     string            `json:"level"`
+	Meta      map[string]string `json:"meta"`
+}
+
+type MezmoLogBody struct {
+	Lines []MezmoLogLine `json:"lines"`
+}
+
+func newLogsExporter(config *Config, settings component.TelemetrySettings) *mezmoExporter {
+	var e = &mezmoExporter{
+		config:   config,
+		settings: settings,
+	}
+	return e
+}
+
+func (m *mezmoExporter) pushLogData(ctx context.Context, ld plog.Logs) error {
+	m.wg.Add(1)
+	defer m.wg.Done()
+
+	return m.logDataToMezmo(ld)
+}
+
+func (m *mezmoExporter) start(_ context.Context, host component.Host) (err error) {
+	client, err := m.config.HTTPClientSettings.ToClient(host.GetExtensions(), m.settings)
+	if err != nil {
+		return err
+	}
+
+	m.client = client
+
+	return nil
+}
+
+func (m *mezmoExporter) stop(context.Context) (err error) {
+	m.wg.Wait()
+	return nil
+}
+
+func (m *mezmoExporter) logDataToMezmo(ld plog.Logs) error {
+	var errs error
+
+	var lines []MezmoLogLine
+
+	// Convert the log resources to mezmo lines...
+	resourceLogs := ld.ResourceLogs()
+	for i := 0; i < resourceLogs.Len(); i++ {
+		ills := resourceLogs.At(i).ScopeLogs()
+		//resource := resourceLogs.At(i).Resource()
+		for j := 0; j < ills.Len(); j++ {
+			logs := ills.At(j).LogRecords()
+
+			for k := 0; k < logs.Len(); k++ {
+				var log = logs.At(k)
+
+				// Convert Attributes to meta fields being mindful of the maxMetaDataSize restriction
+				var attrs = map[string]string{}
+				attrs["trace.id"] = log.TraceID().HexString()
+				attrs["span.id"] = log.SpanID().HexString()
+				log.Attributes().Range(func(k string, v pcommon.Value) bool {
+					attrs[k] = truncateString(v.StringVal(), maxMetaDataSize)
+					return true
+				})
+
+				var s pcommon.Value
+				s, _ = log.Attributes().Get("appname")
+				var app = s.StringVal()
+
+				var line = MezmoLogLine{
+					Timestamp: log.Timestamp().AsTime().UTC().UnixMilli(),
+					Line:      truncateString(log.Body().StringVal(), maxMessageSize),
+					App:       truncateString(app, maxAppnameLen),
+					Level:     truncateString(log.SeverityText(), maxLogLevelLen),
+					Meta:      attrs,
+				}
+				lines = append(lines, line)
+			}
+		}
+	}
+
+	// Send them to Mezmo in batches < 10MB in size
+	var b strings.Builder
+	b.WriteString("{\"lines\": [")
```

Comment:
I don't love writing out json string parts here.  Is there a way to use a json marshaller to accomplish this?

Reply:
I'm not a fan either but given the way we are iterating over the variable length log entries and chunking them up into our JSON structure, I considered this to be the lesser of two evils.  It's a trade off of speed and efficiency vs overhead of using the marshaller and the compexity it creates in the code. The purpose is to accumulate the maximum payload size and when it exceeds 10MB, send what's accumulated and then proceed with the remaining log entries that are available.

Reply:
Got it, thanks.